File: openvidu-call-back/src/controllers/RecordingController.ts
import { CALL_OPENVIDU_CERTTYPE, CALL_RECORDING, OPENVIDU_URL } from '../config';

import * as express from 'express';
import { Request, Response } from 'express';
import { createProxyMiddleware } from 'http-proxy-middleware';
import { Recording } from 'openvidu-node-client';
import { AuthService } from '../services/AuthService';
import { OpenViduService } from '../services/OpenViduService';
export const app = express.Router({
	strict: true
});

const openviduService = OpenViduService.getInstance();
const authService = AuthService.getInstance();

app.get('/', async (req: Request, res: Response) => {
	try {
		const IS_RECORDING_ENABLED = CALL_RECORDING.toUpperCase() === 'ENABLED';
		const sessionId = openviduService.getSessionIdFromCookie(req.cookies);
		const adminSessionId = req.cookies[authService.ADMIN_COOKIE_NAME];
		const isAdminSessionValid = authService.isAdminSessionValid(adminSessionId);
		const isModeratorSessionValid = openviduService.isModeratorSessionValid(sessionId, req.cookies);
		const isParticipantSessionValid = openviduService.isParticipantSessionValid(sessionId, req.cookies);

		if (!IS_RECORDING_ENABLED) {
			const message = 'Recording is disabled';
			console.error(message);
			return res.status(403).json({ message: 'Recording is disabled' });
		}

		if (isAdminSessionValid) {
			const recordings = await openviduService.listAllRecordings();
			return res.status(200).json(recordings);
		}

		if (Boolean(sessionId) && (isModeratorSessionValid || isParticipantSessionValid)) {
			const date = openviduService.getDateFromCookie(req.cookies);
			const recordings = await openviduService.listRecordingsBySessionIdAndDate(sessionId, date);
			return res.status(200).json(recordings);
		}

		const message = 'Permissions denied to drive recording';
		console.error(message);
		return res.status(403).json({ message });
	} catch (error) {
		console.log(error);
		const code = Number(error?.message);
		let message = 'Unexpected error getting all recordings';
		if (code === 404) {
			message = 'No recording exist for the session';
		}
		return res.status(Number(code) || 500).send(JSON.stringify({ message }));
	}
});

app.post('/start', async (req: Request, res: Response) => {
	try {
		let sessionId: string = req.body.sessionId;

		if (!sessionId) {
			return res.status(400).json({ message: 'Missing session id parameter.' });
		}
		if (CALL_RECORDING !== 'ENABLED') {
			const message = 'OpenVidu Call Recording is disabled';
			console.error(`Start recording failed. ${message}`);
			return res.status(403).json({ message });
		}

		if (!openviduService.isModeratorSessionValid(sessionId, req.cookies)) {
			console.log(`Permissions denied for starting recording in session ${sessionId}`);
			return res.status(403).json({ message: 'Permissions denied to drive recording' });
		}

		console.log(`Starting recording in ${sessionId}`);
		const startingRecording = await openviduService.startRecording(sessionId);
		openviduService.moderatorsCookieMap.get(sessionId).recordingId = startingRecording.id;
		return res.status(200).json(startingRecording);
	} catch (error) {
		console.log(error);
		const code = Number(error?.message);
		let message = `Unexpected error starting recording`;
		if (code === 409) {
			message = 'The session is already being recorded.';
		} else if (code === 501) {
			message = 'OpenVidu Server recording module is disabled';
		} else if (code === 406) {
			message = 'The session has no connected participants';
		}
		return res.status(code || 500).send(JSON.stringify({ message }));
	}
});

app.post('/stop', async (req: Request, res: Response) => {
	try {
		let sessionId: string = req.body.sessionId;
		if (!sessionId) {
			return res.status(400).json({ message: 'Missing session id parameter.' });
		}
		if (CALL_RECORDING !== 'ENABLED') {
			const message = 'OpenVidu Call Recording is disabled';
			console.log(`Stop recording failed. ${message}`);
			return res.status(403).json({ message });
		}

		if (!openviduService.isModeratorSessionValid(sessionId, req.cookies)) {
			console.log(`Permissions denied for stopping recording in session ${sessionId}`);
			return res.status(403).json({ message: 'Permissions denied to drive recording' });
		}

		const recordingId = openviduService.moderatorsCookieMap.get(sessionId)?.recordingId;
		if (!recordingId) {
			return res.status(404).json({ message: 'Session was not being recorded' });
		}

		console.log(`Stopping recording in ${sessionId}`);
		await openviduService.stopRecording(recordingId);
		const date = openviduService.getDateFromCookie(req.cookies);
		const recordingList = await openviduService.listRecordingsBySessionIdAndDate(sessionId, date);
		openviduService.moderatorsCookieMap.get(sessionId).recordingId = '';
		res.status(200).json(recordingList);
	} catch (error) {
		console.log(error);
		const code = Number(error?.message);
		let message = `Unexpected error stopping recording`;
		if (code === 501) {
			message = 'OpenVidu Server recording module is disabled';
		} else if (code === 406) {
			message = 'Recording has STARTING status. Wait until STARTED status before stopping the recording';
		}
		return res.status(code || 500).send(JSON.stringify({ message }));
	}
});

app.delete('/delete/:recordingId', async (req: Request, res: Response) => {
	try {
		const recordingId: string = req.params.recordingId;
		if (!recordingId) {
			return res.status(400).json({ message: 'Missing recording id parameter.' });
		}
		const sessionId = openviduService.getSessionIdFromRecordingId(recordingId);
		const adminSessionId = req.cookies[authService.ADMIN_COOKIE_NAME];
		const isModeratorSessionValid = openviduService.isModeratorSessionValid(sessionId, req.cookies);
		const isAdminSessionValid = authService.isAdminSessionValid(adminSessionId);
		let recordings = [];

		if (!(isModeratorSessionValid || isAdminSessionValid)) {
			return res.status(403).json({ message: 'Permissions denied to drive recording' });
		}

		console.log(`Deleting recording ${recordingId}`);
		await openviduService.deleteRecording(recordingId);
		if (isAdminSessionValid) {
			recordings = await openviduService.listAllRecordings();
		} else {
			const date = openviduService.getDateFromCookie(req.cookies);
			recordings = await openviduService.listRecordingsBySessionIdAndDate(sessionId, date);
		}
		res.status(200).json(recordings);
	} catch (error) {
		console.log(error);
		const code = Number(error?.message);
		let message = `Unexpected error deleting the recording`;
		if (code === 409) {
			message = 'The recording has STARTED status. Stop it before deletion.';
		} else if (code === 501) {
			message = 'OpenVidu Server recording module is disabled';
		} else if (code === 409) {
			message = 'No recording exists for the session';
		}
		return res.status(code).send(JSON.stringify({ message }));
	}
});

export const proxyGETRecording = createProxyMiddleware({
	target: `${OPENVIDU_URL}/openvidu/`,
	secure: CALL_OPENVIDU_CERTTYPE !== 'selfsigned',
	onProxyReq: (proxyReq, req: Request, res: Response) => {
		proxyReq.removeHeader('Cookie');
		const recordingId: string = req.params.recordingId;
		if (!recordingId) {
			return res.status(400).send('Missing recording id parameter.');
		}
		const adminSessionId = req.cookies[authService.ADMIN_COOKIE_NAME];
		const isAdminSessionValid = authService.isAdminSessionValid(adminSessionId);
		const sessionId = openviduService.getSessionIdFromRecordingId(recordingId);
		const isModeratorSessionValid = openviduService.isModeratorSessionValid(sessionId, req.cookies);
		const isParticipantSessionValid = openviduService.isParticipantSessionValid(sessionId, req.cookies);

		if (!!sessionId && (isModeratorSessionValid || isParticipantSessionValid || isAdminSessionValid)) {
			proxyReq.setHeader('Connection', 'keep-alive');
			proxyReq.setHeader('Authorization', openviduService.getBasicAuth());
		} else {
			return res.status(403).send(JSON.stringify({ message: 'Permissions denied to drive recording' }));
		}
	},
	onProxyRes: (proxyRes, req: Request, res: Response) => {
		proxyRes.headers['set-cookie'] = null;
	},
	onError: (error, req: Request, res: Response) => {
		console.log(error);
		const code = Number(error?.message);
		let message = 'Unexpected error downloading the recording';
		if (code === 404) {
			message = 'No recording exist for the session';
		}
		res.status(Number(code) || 500).send(JSON.stringify({ message }));
		return res.end();
	}
});

File: openvidu-call-back-java/src/main/java/io/openvidu/call/java/controllers/RecordingController.java
package io.openvidu.call.java.controllers;

import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import io.openvidu.call.java.services.AuthService;
import io.openvidu.call.java.services.OpenViduService;
import io.openvidu.call.java.services.ProxyService;
import io.openvidu.java.client.OpenViduHttpException;
import io.openvidu.java.client.OpenViduJavaClientException;
import io.openvidu.java.client.Recording;

@CrossOrigin(origins = "*")
@RestController
@RequestMapping("/recordings")
public class RecordingController {

	@Value("${CALL_RECORDING}")
	private String CALL_RECORDING;

	@Value("${OPENVIDU_URL}")
	private String OPENVIDU_URL;

	@Autowired
	private OpenViduService openviduService;

	@Autowired
	private AuthService authService;

	@Autowired
	private ProxyService proxyService;

	@GetMapping("")
	public ResponseEntity<?> getRecordings(
			@CookieValue(name = OpenViduService.MODERATOR_TOKEN_NAME, defaultValue = "") String moderatorToken,
			@CookieValue(name = OpenViduService.PARTICIPANT_TOKEN_NAME, defaultValue = "") String participantToken,
			@CookieValue(name = AuthService.ADMIN_COOKIE_NAME, defaultValue = "") String adminToken) {
		try {
			boolean IS_RECORDING_ENABLED = CALL_RECORDING.toUpperCase().equals("ENABLED");
			String sessionId = "";
			if(moderatorToken.isEmpty()){
				sessionId = openviduService.getSessionIdFromCookie(participantToken);
			} else {
				sessionId = openviduService.getSessionIdFromCookie(moderatorToken);
			}
			boolean isModeratorSessionValid = openviduService.isModeratorSessionValid(sessionId, moderatorToken);
			boolean isParticipantSessionValid = openviduService.isParticipantSessionValid(sessionId, participantToken);
			boolean isAdminSessionValid = authService.isAdminSessionValid(adminToken);

			if (!IS_RECORDING_ENABLED) {
			    String message = "Recording is disabled";
			    System.err.println(message);
			    return new ResponseEntity<>(message, HttpStatus.FORBIDDEN);
			}

			if (isAdminSessionValid) {
			    return new ResponseEntity<>(this.openviduService.listAllRecordings(), HttpStatus.OK);
			}

			if (!sessionId.isEmpty() && (isModeratorSessionValid || isParticipantSessionValid)) {
			    String cookie = isParticipantSessionValid ? participantToken : moderatorToken;
			    long date = openviduService.getDateFromCookie(cookie);
			    return new ResponseEntity<>(openviduService.listRecordingsBySessionIdAndDate(sessionId, date), HttpStatus.OK);
			}

			return new ResponseEntity<>("Permissions denied to drive recording", HttpStatus.FORBIDDEN);

		} catch (OpenViduJavaClientException | OpenViduHttpException error) {
			error.printStackTrace();
			int code = Integer.parseInt(error.getMessage());
			String message = "Unexpected error getting all recordings";
			if (code == 404) {
				message = "No recording exist for the session";
			}
			System.err.println(message);
			return new ResponseEntity<>(message, HttpStatus.INTERNAL_SERVER_ERROR);
		}

	}

	@PostMapping("/start")
	public ResponseEntity<?> startRecording(@RequestBody(required = true) Map<String, String> params,
			@CookieValue(name = OpenViduService.MODERATOR_TOKEN_NAME, defaultValue = "") String moderatorToken) {

		try {
			String sessionId = params.get("sessionId");
			if (!CALL_RECORDING.equalsIgnoreCase("ENABLED")) {
				String message = "Start recording failed. OpenVidu Call Recording is disabled";
				System.out.println(message);
				return new ResponseEntity<>(message, HttpStatus.FORBIDDEN);
			}

			if (!openviduService.isModeratorSessionValid(sessionId, moderatorToken)) {
				String message = "Permissions denied for starting recording in session " + sessionId;
				System.out.println(message);
				return new ResponseEntity<>(message, HttpStatus.FORBIDDEN);
			}

			Recording startingRecording = openviduService.startRecording(sessionId);
			openviduService.moderatorsCookieMap.get(sessionId).setRecordingId(startingRecording.getId());
			System.out.println("Starting recording in " + sessionId);
			return new ResponseEntity<>(startingRecording, HttpStatus.OK);

		} catch (OpenViduJavaClientException | OpenViduHttpException error) {
			error.printStackTrace();
			int code = Integer.parseInt(error.getMessage());
			String message = "Unexpected error starting recording";
			if (code == 409) {
				message = "The session is already being recorded.";
			} else if (code == 501) {
				message = "OpenVidu Server recording module is disabled";
			} else if (code == 406) {
				message = "The session has no connected participants";
			}
			System.err.println(message);
			return new ResponseEntity<>(message, HttpStatus.INTERNAL_SERVER_ERROR);
		}
	}

	@PostMapping("/stop")
	public ResponseEntity<?> stopRecording(@RequestBody(required = true) Map<String, String> params,
			@CookieValue(name = OpenViduService.MODERATOR_TOKEN_NAME, defaultValue = "") String moderatorToken) {
		try {
			String sessionId = params.get("sessionId");

			if (!CALL_RECORDING.equalsIgnoreCase("ENABLED")) {
				String message = "Stop recording failed. OpenVidu Call Recording is disabled";
				System.out.println(message);
				return new ResponseEntity<>(message, HttpStatus.FORBIDDEN);
			}

			if (!openviduService.isModeratorSessionValid(sessionId, moderatorToken)) {
				String message = "Permissions denied to drive recording";
				System.err.println(message);
				return new ResponseEntity<>(message, HttpStatus.FORBIDDEN);
			}

			String recordingId = openviduService.moderatorsCookieMap.get(sessionId).getRecordingId();
			if (recordingId.isEmpty()) {
				String message = "Session was not being recorded";
				System.err.println(message);
				return new ResponseEntity<>(message, HttpStatus.NOT_FOUND);
			}

			System.out.println("Stopping recording in " + sessionId);
			openviduService.stopRecording(recordingId);
			long date = openviduService.getDateFromCookie(moderatorToken);
			List<Recording> recordingList = openviduService.listRecordingsBySessionIdAndDate(sessionId, date);
			openviduService.moderatorsCookieMap.get(sessionId).setRecordingId("");
			return new ResponseEntity<>(recordingList, HttpStatus.OK);
		} catch (OpenViduJavaClientException | OpenViduHttpException error) {
			error.printStackTrace();
			int code = Integer.parseInt(error.getMessage());
			String message = "Unexpected error stopping recording";
			if (code == 501) {
				message = "OpenVidu Server recording module is disabled";
			} else if (code == 406) {
				message = "Recording has STARTING status. Wait until STARTED status before stopping the recording";
			}
			System.err.println(message);
			return new ResponseEntity<>(message, HttpStatus.INTERNAL_SERVER_ERROR);
		}
	}

	@DeleteMapping("/delete/{recordingId}")
	public ResponseEntity<?> deleteRecording(@PathVariable String recordingId,
			@CookieValue(name = OpenViduService.MODERATOR_TOKEN_NAME, defaultValue = "") String moderatorToken,
			@CookieValue(name = AuthService.ADMIN_COOKIE_NAME, defaultValue = "") String adminToken) {
		try {

			if (recordingId.isEmpty()) {
				return new ResponseEntity<>("Missing recording id parameter.", HttpStatus.BAD_REQUEST);
			}

			List<Recording> recordings = new ArrayList<Recording>();
			String sessionId = openviduService.getSessionIdFromRecordingId(recordingId);
			boolean isAdminSessionValid = authService.isAdminSessionValid(adminToken);
			boolean isModeratorSessionValid = openviduService.isModeratorSessionValid(sessionId, moderatorToken);

			if(!(isModeratorSessionValid || isAdminSessionValid)) {
				String message = "Permissions denied to drive recording";
				System.err.println(message);
				return new ResponseEntity<>(message, HttpStatus.FORBIDDEN);
			}

			System.out.println("Deleting recording " + recordingId);
			openviduService.deleteRecording(recordingId);
			if (isAdminSessionValid) {
				recordings = openviduService.listAllRecordings();
			} else {
				long date = openviduService.getDateFromCookie(moderatorToken);
				recordings = openviduService.listRecordingsBySessionIdAndDate(sessionId, date);
			}
			return new ResponseEntity<>(recordings, HttpStatus.OK);
		} catch (OpenViduJavaClientException | OpenViduHttpException error) {
			error.printStackTrace();
			int code = Integer.parseInt(error.getMessage());
			String message = "Unexpected error deleting the recording";
			if (code == 409) {
				message = "The recording has STARTED status. Stop it before deletion.";
			} else if (code == 501) {
				message = "OpenVidu Server recording module is disabled";
			} else if (code == 409) {
				message = "No recording exists for the session";
			}
			System.err.println(message);
			return new ResponseEntity<>(message, HttpStatus.INTERNAL_SERVER_ERROR);
		}
	}

	@GetMapping("/{recordingId}/{extension}")
	public ResponseEntity<?> getRecording(@PathVariable String recordingId, @PathVariable String extension,
			@CookieValue(name = OpenViduService.MODERATOR_TOKEN_NAME, defaultValue = "") String moderatorToken,
			@CookieValue(name = OpenViduService.PARTICIPANT_TOKEN_NAME, defaultValue = "") String participantToken,
			@CookieValue(name = AuthService.ADMIN_COOKIE_NAME, defaultValue = "") String sessionToken,
			HttpServletRequest req, HttpServletResponse res) {

		if (recordingId.isEmpty()) {
			return new ResponseEntity<>("Missing recording id parameter.", HttpStatus.BAD_REQUEST);
		}

		String sessionId = this.openviduService.getSessionIdFromRecordingId(recordingId);
		boolean isAdminSessionValid = authService.isAdminSessionValid(sessionToken);
		boolean isModeratorSessionValid = openviduService.isModeratorSessionValid(sessionId, moderatorToken);
		boolean isParticipantSessionValid = openviduService.isParticipantSessionValid(sessionId, participantToken);
		if (!sessionId.isEmpty() && (isModeratorSessionValid || isParticipantSessionValid || isAdminSessionValid)) {

			try {
				return proxyService.recordingProxyRequest(req, res);
			} catch (URISyntaxException e) {
				e.printStackTrace();
			}
			return new ResponseEntity<>(null, HttpStatus.INTERNAL_SERVER_ERROR);

		} else {
			String message = "Permissions denied to drive recording";
			System.err.println(message);
			return new ResponseEntity<>(message, HttpStatus.FORBIDDEN);
		}

	}

}
